File: tracks/alpha/drivers/python/src/scratchbird/metadata.py
"""Metadata query helpers and schema-tree shaping utilities."""

from __future__ import annotations

from dataclasses import dataclass, field
import re
from typing import Dict, Iterable, List, Optional, Set

SCHEMAS_QUERY = (
    "SELECT schema_id, schema_name, owner_id, default_tablespace_id "
    "FROM sys.schemas WHERE is_valid = 1 ORDER BY schema_name"
)
TABLES_QUERY = (
    "SELECT table_id, schema_id, table_name, table_type, owner_id "
    "FROM sys.tables WHERE is_valid = 1 ORDER BY table_name"
)
COLUMNS_QUERY = (
    "SELECT column_id, table_id, column_name, data_type_id, data_type_name, "
    "ordinal_position, is_nullable, default_value, domain_id, collation_id, "
    "charset_id, is_identity, is_generated, generation_expression "
    "FROM sys.columns WHERE is_valid = 1 ORDER BY table_id, ordinal_position"
)
INDEXES_QUERY = (
    "SELECT index_id, table_id, index_name, index_type, is_unique "
    "FROM sys.indexes WHERE is_valid = 1 ORDER BY table_id, index_name"
)
INDEX_COLUMNS_QUERY = (
    "SELECT index_id, column_id, column_name, ordinal_position, is_included "
    "FROM sys.index_columns ORDER BY index_id, ordinal_position"
)
CONSTRAINTS_QUERY = (
    "SELECT constraint_id, table_id, constraint_name, constraint_type "
    "FROM sys.constraints WHERE is_valid = 1 ORDER BY table_id, constraint_name"
)
PROCEDURES_QUERY = (
    "SELECT procedure_id, schema_id, procedure_name, routine_type "
    "FROM sys.procedures WHERE is_valid = 1 ORDER BY schema_id, procedure_name"
)
FUNCTIONS_QUERY = (
    "SELECT function_id, schema_id, function_name "
    "FROM sys.functions WHERE is_valid = 1 ORDER BY schema_id, function_name"
)
ROUTINES_QUERY = (
    "SELECT procedure_id AS routine_id, schema_id, procedure_name AS routine_name, routine_type "
    "FROM sys.procedures WHERE is_valid = 1 "
    "UNION ALL "
    "SELECT function_id AS routine_id, schema_id, function_name AS routine_name, 'FUNCTION' AS routine_type "
    "FROM sys.functions WHERE is_valid = 1 "
    "ORDER BY schema_id, routine_name"
)
CATALOGS_QUERY = (
    "SELECT schema_id AS catalog_id, schema_name AS catalog_name "
    "FROM sys.schemas WHERE is_valid = 1 ORDER BY schema_name"
)
PRIMARY_KEYS_QUERY = (
    "SELECT constraint_id, table_id, constraint_name, constraint_type "
    "FROM sys.constraints WHERE is_valid = 1 "
    "AND lower(constraint_type) IN ('primary key', 'primary') "
    "ORDER BY table_id, constraint_name"
)
FOREIGN_KEYS_QUERY = (
    "SELECT constraint_id, table_id, constraint_name, constraint_type "
    "FROM sys.constraints WHERE is_valid = 1 "
    "AND lower(constraint_type) IN ('foreign key', 'foreign') "
    "ORDER BY table_id, constraint_name"
)
TABLE_PRIVILEGES_QUERY = (
    "SELECT table_id, table_name, owner_id AS grantor_id, owner_id AS grantee_id, "
    "'ALL' AS privilege_type "
    "FROM sys.tables WHERE is_valid = 1 ORDER BY table_id, table_name"
)
COLUMN_PRIVILEGES_QUERY = (
    "SELECT table_id, column_id, column_name, 'ALL' AS privilege_type "
    "FROM sys.columns WHERE is_valid = 1 ORDER BY table_id, ordinal_position"
)
TYPE_INFO_QUERY = (
    "SELECT DISTINCT data_type_id, data_type_name "
    "FROM sys.columns WHERE is_valid = 1 ORDER BY data_type_name"
)

DEFAULT_COLLECTION = "tables"

COLLECTION_QUERY_MAP: Dict[str, str] = {
    "schemas": SCHEMAS_QUERY,
    "tables": TABLES_QUERY,
    "columns": COLUMNS_QUERY,
    "indexes": INDEXES_QUERY,
    "index_columns": INDEX_COLUMNS_QUERY,
    "constraints": CONSTRAINTS_QUERY,
    "procedures": PROCEDURES_QUERY,
    "functions": FUNCTIONS_QUERY,
    "routines": ROUTINES_QUERY,
    "catalogs": CATALOGS_QUERY,
    "primary_keys": PRIMARY_KEYS_QUERY,
    "foreign_keys": FOREIGN_KEYS_QUERY,
    "table_privileges": TABLE_PRIVILEGES_QUERY,
    "column_privileges": COLUMN_PRIVILEGES_QUERY,
    "type_info": TYPE_INFO_QUERY,
}

COLLECTION_ALIASES: Dict[str, str] = {
    "schema": "schemas",
    "schemas": "schemas",
    "table": "tables",
    "tables": "tables",
    "column": "columns",
    "columns": "columns",
    "index": "indexes",
    "indexes": "indexes",
    "index_column": "index_columns",
    "index_columns": "index_columns",
    "indexcolumn": "index_columns",
    "indexcolumns": "index_columns",
    "constraint": "constraints",
    "constraints": "constraints",
    "procedure": "procedures",
    "procedures": "procedures",
    "function": "functions",
    "functions": "functions",
    "routine": "routines",
    "routines": "routines",
    "catalog": "catalogs",
    "catalogs": "catalogs",
    "primary_key": "primary_keys",
    "primary_keys": "primary_keys",
    "primarykey": "primary_keys",
    "primarykeys": "primary_keys",
    "foreign_key": "foreign_keys",
    "foreign_keys": "foreign_keys",
    "foreignkey": "foreign_keys",
    "foreignkeys": "foreign_keys",
    "table_privilege": "table_privileges",
    "table_privileges": "table_privileges",
    "tableprivilege": "table_privileges",
    "tableprivileges": "table_privileges",
    "column_privilege": "column_privileges",
    "column_privileges": "column_privileges",
    "columnprivilege": "column_privileges",
    "columnprivileges": "column_privileges",
    "type_info": "type_info",
    "typeinfo": "type_info",
}


@dataclass
class SchemaTreeNode:
    """Metadata-only schema tree node for recursive navigation surfaces."""

    name: str
    full_path: str
    is_terminal: bool = False
    children: List["SchemaTreeNode"] = field(default_factory=list)


def schemas_query() -> str:
    return SCHEMAS_QUERY


def tables_query() -> str:
    return TABLES_QUERY


def columns_query() -> str:
    return COLUMNS_QUERY


def indexes_query() -> str:
    return INDEXES_QUERY


def index_columns_query() -> str:
    return INDEX_COLUMNS_QUERY


def constraints_query() -> str:
    return CONSTRAINTS_QUERY


def procedures_query() -> str:
    return PROCEDURES_QUERY


def functions_query() -> str:
    return FUNCTIONS_QUERY


def routines_query() -> str:
    return ROUTINES_QUERY


def catalogs_query() -> str:
    return CATALOGS_QUERY


def primary_keys_query() -> str:
    return PRIMARY_KEYS_QUERY


def foreign_keys_query() -> str:
    return FOREIGN_KEYS_QUERY


def table_privileges_query() -> str:
    return TABLE_PRIVILEGES_QUERY


def column_privileges_query() -> str:
    return COLUMN_PRIVILEGES_QUERY


def type_info_query() -> str:
    return TYPE_INFO_QUERY


def normalize_collection_name(collection_name: Optional[str] = None) -> str:
    raw = DEFAULT_COLLECTION if collection_name is None else str(collection_name)
    normalized = raw.strip().lower().replace("-", "_").replace(" ", "_")
    if not normalized:
        normalized = DEFAULT_COLLECTION
    collapsed = normalized.replace("_", "")
    resolved = COLLECTION_ALIASES.get(normalized) or COLLECTION_ALIASES.get(collapsed)
    if resolved is None:
        raise ValueError(f"metadata collection '{raw}' is not supported")
    return resolved


def resolve_collection_query(collection_name: Optional[str] = None) -> str:
    resolved = normalize_collection_name(collection_name)
    return COLLECTION_QUERY_MAP[resolved]


def schema_name_matches_pattern(schema_name: Optional[str], schema_pattern: Optional[str]) -> bool:
    """Return True when a schema name matches JDBC-style `%`/`_` wildcard pattern."""
    if not schema_pattern:
        return True
    if schema_name is None:
        return False
    return bool(_pattern_to_regex(schema_pattern).match(schema_name))


def schema_paths_for_navigation(
    schema_names: Iterable[str],
    *,
    expand_schema_parents: bool = False,
    schema_pattern: Optional[str] = None,
) -> List[str]:
    """
    Normalize, de-duplicate, and filter schema paths for metadata navigation.

    When `expand_schema_parents` is True, dotted parent segments are emitted in
    insertion order (for example `users`, `users.alice`, `users.alice.dev`).
    """
    if expand_schema_parents:
        return expand_schema_parent_paths(schema_names, schema_pattern=schema_pattern)

    out: List[str] = []
    seen: Set[str] = set()
    for schema_name in schema_names:
        normalized = _normalize_schema_name(schema_name)
        if normalized is None:
            continue
        if not schema_name_matches_pattern(normalized, schema_pattern):
            continue
        if normalized in seen:
            continue
        seen.add(normalized)
        out.append(normalized)
    return out


def expand_schema_parent_paths(
    schema_names: Iterable[str],
    *,
    schema_pattern: Optional[str] = None,
) -> List[str]:
    """
    Expand dotted schema names to include parent segments.

    Behavior mirrors the JDBC metadata expansion mode:
    `users.alice.dev` -> `users`, `users.alice`, `users.alice.dev`.
    """
    out: List[str] = []
    seen: Set[str] = set()
    for schema_name in schema_names:
        parts = _split_schema_path(schema_name)
        if not parts:
            continue
        current: List[str] = []
        for part in parts:
            current.append(part)
            candidate = ".".join(current)
            if not schema_name_matches_pattern(candidate, schema_pattern):
                continue
            if candidate in seen:
                continue
            seen.add(candidate)
            out.append(candidate)
    return out


def build_schema_tree(schema_paths: Iterable[str]) -> List[SchemaTreeNode]:
    """
    Build a metadata-only recursive schema tree from dotted schema paths.

    - Child name uniqueness is enforced per parent.
    - Same-name nodes in different schema paths are preserved as distinct nodes.
    """
    nodes_by_path: Dict[str, SchemaTreeNode] = {}
    roots: List[SchemaTreeNode] = []

    for schema_path in schema_paths:
        parts = _split_schema_path(schema_path)
        if not parts:
            continue

        parent: Optional[SchemaTreeNode] = None
        current_path: List[str] = []
        for part in parts:
            current_path.append(part)
            full_path = ".".join(current_path)
            node = nodes_by_path.get(full_path)
            if node is None:
                node = SchemaTreeNode(name=part, full_path=full_path)
                nodes_by_path[full_path] = node
                if parent is None:
                    roots.append(node)
                else:
                    parent.children.append(node)
            parent = node

        if parent is not None:
            parent.is_terminal = True

    return roots


def _normalize_schema_name(schema_name: Optional[str]) -> Optional[str]:
    parts = _split_schema_path(schema_name)
    if not parts:
        return None
    return ".".join(parts)


def _split_schema_path(schema_name: Optional[str]) -> List[str]:
    if schema_name is None:
        return []
    parts: List[str] = []
    for part in str(schema_name).split("."):
        normalized = part.strip()
        if normalized:
            parts.append(normalized)
    return parts


def _pattern_to_regex(pattern: str) -> re.Pattern[str]:
    out = ["^"]
    escaped = False
    for ch in pattern:
        if escaped:
            out.append(re.escape(ch))
            escaped = False
            continue
        if ch == "\\":
            escaped = True
            continue
        if ch == "%":
            out.append(".*")
        elif ch == "_":
            out.append(".")
        else:
            out.append(re.escape(ch))
    out.append("$")
    return re.compile("".join(out), re.IGNORECASE)

File: tracks/alpha/drivers/ruby/lib/scratchbird/metadata.rb
# ScratchBird-driver
# Copyright (c) 2025-2026 Dalton Calford
#
# Licensed under the Initial Developer's Public License Version 1.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at:
# https://www.firebirdsql.org/en/initial-developer-s-public-license-version-1-0/

module Scratchbird
  module Metadata
    SCHEMAS_QUERY = "SELECT schema_id, schema_name, owner_id, default_tablespace_id FROM sys.schemas WHERE is_valid = 1 ORDER BY schema_name"
    TABLES_QUERY = "SELECT table_id, schema_id, table_name, table_type, owner_id FROM sys.tables WHERE is_valid = 1 ORDER BY table_name"
    COLUMNS_QUERY = "SELECT column_id, table_id, column_name, data_type_id, data_type_name, ordinal_position, is_nullable, default_value, domain_id, collation_id, charset_id, is_identity, is_generated, generation_expression FROM sys.columns WHERE is_valid = 1 ORDER BY table_id, ordinal_position"
    INDEXES_QUERY = "SELECT index_id, table_id, index_name, index_type, is_unique FROM sys.indexes WHERE is_valid = 1 ORDER BY table_id, index_name"
    INDEX_COLUMNS_QUERY = "SELECT index_id, column_id, column_name, ordinal_position, is_included FROM sys.index_columns ORDER BY index_id, ordinal_position"
    CONSTRAINTS_QUERY = "SELECT constraint_id, table_id, constraint_name, constraint_type FROM sys.constraints WHERE is_valid = 1 ORDER BY table_id, constraint_name"
    PROCEDURES_QUERY = "SELECT procedure_id, schema_id, procedure_name, routine_type FROM sys.procedures WHERE is_valid = 1 ORDER BY schema_id, procedure_name"
    FUNCTIONS_QUERY = "SELECT function_id, schema_id, function_name FROM sys.functions WHERE is_valid = 1 ORDER BY schema_id, function_name"
    SCHEMA_FIELD_CANDIDATES = %w[
      schema_name
      table_schem
      table_schema
      schema
      SCHEMA_NAME
      TABLE_SCHEM
      TABLE_SCHEMA
      SCHEMA
    ].freeze

    class SchemaTreeNode
      attr_reader :name, :full_path, :children
      attr_accessor :terminal

      def initialize(name, full_path)
        @name = name
        @full_path = full_path
        @terminal = false
        @children = []
      end
    end

    def self.schemas_query
      SCHEMAS_QUERY
    end

    def self.tables_query
      TABLES_QUERY
    end

    def self.columns_query
      COLUMNS_QUERY
    end

    def self.indexes_query
      INDEXES_QUERY
    end

    def self.index_columns_query
      INDEX_COLUMNS_QUERY
    end

    def self.constraints_query
      CONSTRAINTS_QUERY
    end

    def self.procedures_query
      PROCEDURES_QUERY
    end

    def self.functions_query
      FUNCTIONS_QUERY
    end

    def self.schema_paths_for_navigation(rows_or_names, expand_schema_parents: false)
      base_paths = unique_schema_paths(rows_or_names)
      return base_paths unless expand_schema_parents

      expand_schema_parent_paths(base_paths)
    end

    def self.expand_schema_parent_paths(rows_or_names)
      out = []
      seen = {}

      each_schema_path(rows_or_names) do |schema_path|
        current = +""
        split_schema_path(schema_path).each do |segment|
          current = current.empty? ? segment : "#{current}.#{segment}"
          next if seen[current]

          seen[current] = true
          out << current
        end
      end

      out
    end

    def self.build_schema_tree(schema_paths)
      nodes_by_path = {}
      roots = []

      each_schema_path(schema_paths) do |schema_path|
        parent = nil
        path_parts = []

        split_schema_path(schema_path).each do |segment|
          path_parts << segment
          full_path = path_parts.join(".")

          node = nodes_by_path[full_path]
          unless node
            node = SchemaTreeNode.new(segment, full_path)
            nodes_by_path[full_path] = node
            if parent
              parent.children << node
            else
              roots << node
            end
          end

          parent = node
        end

        parent.terminal = true if parent
      end

      roots
    end

    def self.expand_schema_metadata_rows(rows)
      out = []
      seen = {}

      rows.each do |row|
        schema_path = read_schema_path(row)
        unless schema_path
          out << row
          next
        end

        current = +""
        segments = split_schema_path(schema_path)
        segments.each_with_index do |segment, idx|
          current = current.empty? ? segment : "#{current}.#{segment}"
          next if seen[current]

          seen[current] = true
          if idx == segments.length - 1
            out << row
          else
            out << synthetic_schema_row(row, current)
          end
        end
      end

      out
    end

    def self.build_database_default_metadata_rows(rows_or_names, database:, expand_schema_parents: false, default_branch: "default")
      database_name = database.to_s.strip
      raise ArgumentError, "database is required" if database_name.empty?

      default_name = default_branch.to_s.strip
      raise ArgumentError, "default_branch is required" if default_name.empty?

      schema_paths = schema_paths_for_navigation(rows_or_names, expand_schema_parents: expand_schema_parents)
      roots = build_schema_tree(schema_paths)
      rows = []

      database_path = database_name
      default_path = "#{database_path}.#{default_name}"

      rows << {
        "node_type" => "database",
        "node_name" => database_name,
        "node_path" => database_path,
        "parent_path" => nil,
        "schema_path" => nil,
        "terminal" => false
      }

      rows << {
        "node_type" => "schema",
        "node_name" => default_name,
        "node_path" => default_path,
        "parent_path" => database_path,
        "schema_path" => nil,
        "terminal" => false
      }

      append_tree_metadata_rows(rows, roots, default_path)
      rows
    end

    def self.append_tree_metadata_rows(out_rows, nodes, parent_node_path)
      nodes.each do |node|
        node_path = "#{parent_node_path}.#{node.name}"
        out_rows << {
          "node_type" => "schema",
          "node_name" => node.name,
          "node_path" => node_path,
          "parent_path" => parent_node_path,
          "schema_path" => node.full_path,
          "terminal" => node.terminal
        }
        append_tree_metadata_rows(out_rows, node.children, node_path)
      end
    end
    private_class_method :append_tree_metadata_rows

    def self.unique_schema_paths(rows_or_names)
      out = []
      seen = {}

      each_schema_path(rows_or_names) do |schema_path|
        next if seen[schema_path]

        seen[schema_path] = true
        out << schema_path
      end

      out
    end
    private_class_method :unique_schema_paths

    def self.each_schema_path(rows_or_names)
      return enum_for(:each_schema_path, rows_or_names) unless block_given?
      return if rows_or_names.nil?

      enumerated =
        if rows_or_names.is_a?(String) || rows_or_names.is_a?(Hash)
          [rows_or_names]
        elsif rows_or_names.respond_to?(:each)
          rows_or_names
        else
          [rows_or_names]
        end

      enumerated.each do |row_or_name|
        schema_path = read_schema_path(row_or_name)
        next unless schema_path

        yield schema_path
      end
    end
    private_class_method :each_schema_path

    def self.read_schema_path(row_or_name)
      return normalize_schema_path(row_or_name) if row_or_name.is_a?(String)

      return normalize_schema_path(row_or_name.to_s) if row_or_name.is_a?(Symbol)

      return read_schema_path_from_hash(row_or_name) if row_or_name.is_a?(Hash)

      nil
    end
    private_class_method :read_schema_path

    def self.read_schema_path_from_hash(row)
      SCHEMA_FIELD_CANDIDATES.each do |candidate|
        value = nil
        value = row[candidate] if row.key?(candidate)
        symbol = candidate.to_sym
        value = row[symbol] if value.nil? && row.key?(symbol)
        next unless value.is_a?(String)

        normalized = normalize_schema_path(value)
        return normalized if normalized
      end

      nil
    end
    private_class_method :read_schema_path_from_hash

    def self.normalize_schema_path(value)
      normalized = split_schema_path(value).join(".")
      return nil if normalized.empty?

      normalized
    end
    private_class_method :normalize_schema_path

    def self.split_schema_path(value)
      value.to_s.split(".").map(&:strip).reject(&:empty?)
    end
    private_class_method :split_schema_path

    def self.synthetic_schema_row(sample_row, schema_path)
      synthetic = {}
      sample_row.each_key do |key|
        synthetic[key] = nil
      end
      assign_schema_path(synthetic, schema_path)
      synthetic
    end
    private_class_method :synthetic_schema_row

    def self.assign_schema_path(row, schema_path)
      assigned = false
      SCHEMA_FIELD_CANDIDATES.each do |candidate|
        if row.key?(candidate)
          row[candidate] = schema_path
          assigned = true
        end
        symbol = candidate.to_sym
        if row.key?(symbol)
          row[symbol] = schema_path
          assigned = true
        end
      end
      return if assigned

      target_key = row.keys.any? { |key| key.is_a?(Symbol) } ? :schema_name : "schema_name"
      row[target_key] = schema_path
    end
    private_class_method :assign_schema_path
  end
end
